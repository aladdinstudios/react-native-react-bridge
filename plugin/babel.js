const { parseSync, transformSync } = require("@babel/core");
const traverse = require("@babel/traverse").default;
const p = require("path");

export const extractEntryPaths = (src, filename) => {
  const ast = parseSync(src, { filename });
  const compPaths = [];
  traverse(ast, {
    CallExpression(path) {
      if (isBundleReact(path.node)) {
        const [source, ...args] = path.get("arguments");
        compPaths.push(source.node.value);
      }
    },
  });
  return compPaths;
};

export const replaceEntryPaths = (src, filename, compStrs) => {
  return transformSync(src, {
    filename,
    // babelrc: false,
    plugins: [
      {
        visitor: {
          CallExpression(path, { file: { opts: fileOpts } }) {
            if (isBundleReact(path.node)) {
              const [source, ...args] = path.get("arguments");
              path.replaceWithSourceString(compStrs[source.node.value]);
            }
          },
        },
      },
    ],
  });
};

function isBundleReact(node) {
  return (
    looksLike(node, {
      callee: {
        type: "Identifier",
        name: "bundleReact",
      },
    }) ||
    looksLike(node, {
      callee: {
        type: "MemberExpression",
        property: { name: "bundleReact" },
      },
    })
  );
}

function looksLike(a, b) {
  return (
    a &&
    b &&
    Object.keys(b).every((bKey) => {
      const bVal = b[bKey];
      const aVal = a[bKey];
      if (typeof bVal === "function") {
        return bVal(aVal);
      }
      return isPrimitive(bVal) ? bVal === aVal : looksLike(aVal, bVal);
    })
  );
}

function isPrimitive(val) {
  return val == null || /^[sbn]/.test(typeof val);
}
