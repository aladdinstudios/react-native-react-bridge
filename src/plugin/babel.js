import { parseSync, transformSync } from "@babel/core";
import traverse from "@babel/traverse";

export const isEntryFile = (src, filename) => {
  const ast = parseSync(src, { filename });
  let isEntry = false;
  traverse(ast, {
    ExportDefaultDeclaration(path) {
      if (
        looksLike(path.node, {
          declaration: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "webViewRender",
            },
          },
        })
      ) {
        isEntry = true;
      }
    },
  });
  return isEntry;
};

function looksLike(a, b) {
  return (
    a &&
    b &&
    Object.keys(b).every((bKey) => {
      const bVal = b[bKey];
      const aVal = a[bKey];
      if (typeof bVal === "function") {
        return bVal(aVal);
      }
      return isPrimitive(bVal) ? bVal === aVal : looksLike(aVal, bVal);
    })
  );
}

function isPrimitive(val) {
  return val == null || /^[sbn]/.test(typeof val);
}
