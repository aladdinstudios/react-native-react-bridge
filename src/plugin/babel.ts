import { parseSync } from "@babel/core";
import traverse from "@babel/traverse";

/**
 * @internal
 */
export const isEntryFile = (src: string, filename: string) => {
  // TODO try lighter approach
  const ast = parseSync(src, { filename })!;
  let isEntry = false;
  traverse(ast, {
    ExportDefaultDeclaration(path) {
      if (
        looksLike(path.node, {
          declaration: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "webViewRender",
            },
          },
        }) ||
        looksLike(path.node, {
          declaration: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "webViewCreateRoot",
            },
          },
        })
      ) {
        isEntry = true;
      }
    },
  });
  return isEntry;
};

function looksLike(a: any, b: any): boolean {
  return (
    a &&
    b &&
    Object.keys(b).every((bKey) => {
      const bVal = b[bKey as keyof typeof b];
      const aVal = a[bKey as keyof typeof b];
      if (typeof bVal === "function") {
        return bVal(aVal);
      }
      return isPrimitive(bVal) ? bVal === aVal : looksLike(aVal, bVal);
    })
  );
}

function isPrimitive(val: any): boolean {
  return val == null || /^[sbn]/.test(typeof val);
}
