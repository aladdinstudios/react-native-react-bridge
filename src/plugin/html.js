import { ROOT_ID } from "../common";

export const createContent = (js) =>
  `export default String.raw\`${wrapByHtml(js, "")}\`;`;

const wrapByHtml = (js, css) => `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
  ${css && `<style type="text/css">${css}</style>`}
</head>
<body style="margin: 0 !important;padding: 0 !important;">
  <div id="${ROOT_ID}"></div>
  <script type="text/javascript">(function(){${js}})()</script>
</body>
</html>
`;
