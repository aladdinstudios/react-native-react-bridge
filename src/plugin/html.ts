import { ROOT_ID } from "../constants";

/**
 * @internal
 */
export const buildWebEntryModule = (js: string): string => {
  // https://github.com/inokawa/react-native-react-bridge/pull/133
  js = js.replace(/([`$])/g, "\\$1");
  return (
    "export default String.raw`\n" +
    wrapByHtml(js) +
    "\n`.replace(/\\\\([`$])/g, '\\$1')"
  );
};

const wrapByHtml = (js: string): string => `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
</head>
<body style="margin: 0 !important;padding: 0 !important;">
  <div id="${ROOT_ID}"></div>
  <script type="text/javascript">(function(){${js}})()</script>
</body>
</html>
`;
