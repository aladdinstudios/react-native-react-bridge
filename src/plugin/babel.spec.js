import * as fs from "fs";
import * as path from "path";
import { isEntryFile } from "./babel";
import { createContent } from "./html";

const readFixture = (filename) =>
  fs.readFileSync(path.join(__dirname, "../../fixtures", filename), "utf-8");

describe("isEntryFile", () => {
  it("webViewRender(App)", () => {
    const filename = "app-export-non.jsx";
    const src = readFixture(filename);
    const paths = isEntryFile(src, filename);
    expect(paths).toEqual(false);
  });

  it("export default webViewRender(App)", () => {
    const filename = "app-export-default.jsx";
    const src = readFixture(filename);
    const paths = isEntryFile(src, filename);
    expect(paths).toEqual(true);
  });

  it("export const foo = webViewRender(App)", () => {
    const filename = "app-export-named.jsx";
    const src = readFixture(filename);
    const paths = isEntryFile(src, filename);
    expect(paths).toEqual(false);
  });

  it("module.exports = webViewRender(App)", () => {
    const filename = "app-export-commonjs.jsx";
    const src = readFixture(filename);
    const paths = isEntryFile(src, filename);
    expect(paths).toEqual(false);
  });
});

describe("createContent", () => {
  it("export default webViewRender(App)", () => {
    const filename = "app-export-default.jsx";
    const src = readFixture(filename);
    expect(createContent(src)).toMatchSnapshot();
  });
});
