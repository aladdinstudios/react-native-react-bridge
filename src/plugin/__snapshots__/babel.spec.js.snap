// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContent export default webViewRender(App) 1`] = `
"export default String.raw\`
<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"utf-8\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no\\" />
  
</head>
<body style=\\"margin: 0 !important;padding: 0 !important;\\">
  <div id=\\"root\\"></div>
  <script type=\\"text/javascript\\">(function(){import React from \\"react\\";
import { webViewRender } from \\"react-native-react-bridge/lib/web\\";
import Comp from \\"./Component\\";
import './example.css'

const App = () => {
  return <Comp />;
};

export default webViewRender(App);
})()</script>
</body>
</html>
\`;"
`;
